homeassistant/components/hue/sensor.py

```diff
@@ -0,0 +1,324 @@
+    import aiohue
+
+    api = bridge.api.sensors
+
+    try:
+        start = monotonic()
+        with async_timeout.timeout(4):
+            await api.update()
+    except (asyncio.TimeoutError, aiohue.AiohueException) as err:
+        _LOGGER.debug('Failed to fetch sensor: %s', err)
+
+        if not bridge.available:
+            return
+
+        _LOGGER.error('Unable to reach bridge %s (%s)', bridge.host, err)
+        bridge.available = False
+
+        for sensor_id, sensor in current.items():
+            if sensor_id not in progress_waiting:
+                sensor.async_schedule_update_ha_state()
+
+        return
+
+    finally:
+        _LOGGER.debug('Finished sensor request in %.3f seconds',
+                      monotonic() - start)
+
+    if not bridge.available:
+        _LOGGER.info('Reconnected to bridge %s', bridge.host)
+        bridge.available = True
+
+    new_sensors = []
+    sensor_device_names = {}
+
+    for item_id in api:
+        if item_id not in current:
+            name = PRESENCE_NAME_FORMAT.format(api[item_id].name)
+            if api[item_id].type == aiohue.sensors.TYPE_ZLL_PRESENCE:
+                s = HuePresence(
+                    hass, api[item_id], name, request_bridge_update,
+                    bridge)
+                sensor_device_names[s.device_id] = api[item_id].name
+                current[item_id] = s
+
+    # Iterate again now we have all the presence sensors, and add the related
+    # sensors with nice names
+    for item_id in api:
+        if item_id not in current:
+            device_id = api[item_id].uniqueid
+            if device_id and len(device_id) > 23:
+                device_id = device_id[:23]
+            name = api[item_id].name
+            if api[item_id].type == aiohue.sensors.TYPE_ZLL_LIGHTLEVEL:
+                if device_id in sensor_device_names:
+                    name = LIGHT_LEVEL_NAME_FORMAT.format(
+                        sensor_device_names[device_id])
+                current[item_id] = HueLightLevel(
+                    hass, api[item_id], name, request_bridge_update, bridge)
+            elif api[item_id].type == aiohue.sensors.TYPE_ZLL_TEMPERATURE:
+                if device_id in sensor_device_names:
+                    name = TEMPERATURE_NAME_FORMAT.format(
+                        sensor_device_names[device_id])
+                current[item_id] = HueTemperature(
+                    hass, api[item_id], name, request_bridge_update, bridge)
+
+            if item_id in current:
+                new_sensors.append(current[item_id])
+
+        elif item_id not in progress_waiting:
+            current[item_id].async_schedule_update_ha_state()
+
+    if new_sensors:
+        async_add_entities(new_sensors)
+
+
+class GenericHueSensor:
+    """Representation of a Hue sensor."""
+
+    def __init__(self, hass, sensor, name, request_bridge_update, bridge):
+        """Initialize the sensor."""
+        self.hass = hass
```

Comment:
There is no need for this either.

Reply:
I found it necessary to add this to support `entity_id` generation, so the generator can inspect the list of current entities.

Reply:
Yeah, so let's not  do that 👍 It's set on the entity just before we call `async_added_to_hass`.